### PR TITLE
Recovery fixes

### PR DESCRIPTION
**Pull request by levonpetrosyan93**
### **What issues fixed the PR**
1. Public balance recovery issue. When restoring it is not getting full full public balance as there are unused keys in keypool and it is not trying to create new keys. Default keypool size is 100, and if for some reason user leaves some addresses from the middle unused, it is not recovering balance correct. By setting bigger keypool size it is recovering full balance. 
2. When you try to create a mint, it updates hd counter in memory, and when you broadcast it updates counter in db. So when you cancel the mint transaction the counter remains in memory incremented, and after creating new mint there remains unused counter. Due to this recovery process becomes slower, also during restoration there happens similar thing as for public balance. it stops creating new mintpool as last  index in mintpool is unused. 
3. Due to issue number 2 there are wallets with private balance recovery issues.

**### Code changes.**
1. When we see a transaction we are setting all keys, up to current one used in tx, as used and doing TopUpKeypool() which is creating new keys, Code mainly taken from bitcoin [095142d1f93f39ad2b88dbe8d40140223a1b3900](https://github.com/bitcoin/bitcoin/commit/095142d1f93f39ad2b88dbe8d40140223a1b3900)
2. To fix issue number 2 we are resetting counters before each mint creation, changes are here 163a39ad5d2d9e253d45534611560cca63f4ed92
3. To fix number 3 we are creating one more extra pool and if we see at least one used index we are iterating again. ae8563b5589edf2e48658f7e3f1962f068308a2b
4. Also we do revert of this commit 52b3dd0b06ff313bd1858bc6d1f42087edbb3cf2 and just gave priority to lelanuts during restoration. d67dbe68f8e6e7962594fd4743f034aab326918c

### CODE REVIEW

File: src/hdmint/wallet.cpp [hunk 1]
```diff
@@ -249,19 +251,21 @@ void CHDMintWallet::SetWalletTransactionBlock(CWalletTx &wtx, const CBlockIndex
...
-        foundSigma = false;
-        foundLela = false;
+    int mintsFound = 1;
+    bool firstIteration = true;
+    while (found || mintsFound > 0) {
+        found = false;
+        mintsFound = 0;
         if (fGenerateMintPool)
-            GenerateMintPool(walletdb);
+            GenerateMintPool(walletdb, !firstIteration);
```
Comment (psolstice) [inline src/hdmint/wallet.cpp:264]: Please clarify why the first iteration is different

**Review by psolstice — CHANGES_REQUESTED** [review]

File: src/hdmint/wallet.cpp [hunk 2]
```diff
@@ -248,19 +250,21 @@ void CHDMintWallet::SetWalletTransactionBlock(CWalletTx &wtx, const CBlockIndex
...
+    bool found = true;
 
     std::set<uint256> setAddedTx;
     std::set<uint256> setChecked;
-    while (foundSigma || foundLela) {
-        foundSigma = false;
-        foundLela = false;
+    int mintsFound = 1;
+    bool firstIteration = true;
+    while (found || mintsFound > 0) {
```
[the remark this answers is not in the document]
Comment (levonpetrosyan93) [reply src/hdmint/wallet.cpp:260]: agree

File: src/hdmint/wallet.cpp [hunk 3]
```diff
@@ -277,49 +281,62 @@ void CHDMintWallet::SyncWithChain(bool fGenerateMintPool, boost::optional<std::l
...
+
+            uint160 seedId = std::get<1>(pMint.second);
+            CDataStream ss(SER_GETHASH, 0);
+            ss << pMint.first;
+            ss << seedId;
+            uint256 mintTag = Hash(ss.begin(), ss.end());
+
             COutPoint outPoint;
-            if (!tracker.HasSigmaPubcoinHash(pMint.first, walletdb) && sigma::GetOutPoint(outPoint, pMint.first)) {
+            if (!pwalletMain->IsLocked() && lelantus::GetOutPointFromMintTag(outPoint, mintTag)) {
```
[the remark this answers is not in the document]
Comment (levonpetrosyan93) [reply src/hdmint/wallet.cpp:295]: In Lelantus, you need to derive AES key form master key to decrypt amout of jmint, so if wallet is locked we are not doing the sync for lelantus, we are doing it after unlocked.

File: src/hdmint/wallet.cpp (hunk 1 above)
Comment (levonpetrosyan93) [reply src/hdmint/wallet.cpp:264]: 1. we need to generate new mintpool when we find out that we have used entry in current one, 
2. in each startup we are running this function to update our mint state, 
3 we are preventing generation at first iteration as we dont know the answer of 1, so we are running sync process for current mintpool and then if we find used mints, we are generating a new one

File: src/hdmint/wallet.cpp [hunk 4]
```diff
@@ -277,49 +281,62 @@ void CHDMintWallet::SyncWithChain(bool fGenerateMintPool, boost::optional<std::l
             int32_t& mintCount = std::get<2>(pMint.second);
 
             // halt processing if mint already in tracker
+            if (tracker.HasPubcoinHash(pMint.first, walletdb))
+                continue;
```
[the remark this answers is not in the document]
Comment (levonpetrosyan93) [reply src/hdmint/wallet.cpp:286]: Sorry can you describe your idea more detailed, where we need break?

[the remark this answers is not in the document]
Comment (psolstice) [reply src/hdmint/wallet.cpp:286]: it's functionally equivalent to break. And break is clearer. A lot clerer.

[the remark this answers is not in the document]
Comment (psolstice) [reply src/hdmint/wallet.cpp:286]: there was a typo in the original comment. not = now

[the remark this answers is not in the document]
Comment (levonpetrosyan93) [reply src/hdmint/wallet.cpp:286]: sorry but not agree, `break` will stop `for` loop, but not the `while` witch  takes in account `found` and `mintsFound` , here we just see that we already have the mint in our db, just go onto next mintpool entry, we need to iterate on whole mintpool , then if conditions in `while` are `true` we just create a new `mintpool` and iterate on that.

**Review by psolstice — APPROVED** [review]

**Pull request merged**